neuromation/api/tracing.py

```diff
@@ -0,0 +1,42 @@
+# Distributed tracing support
+
+import binascii
+import os
+import types
+
+import aiohttp
+from multidict import CIMultiDict
+
+
+def gen_trace_id() -> str:
+    """Returns a 64 bit UTF-8 encoded string.
+
+    The id is used for distributed tracing.
+    """
+    return str(binascii.hexlify(os.urandom(8)).decode("utf-8"))
+
+
+def _update_headers(headers: "CIMultiDict[str]", trace_id: str, span_id: str) -> None:
+    """Creates dict with zipkin single header format.
+    """
+    # b3={TraceId}-{SpanId}-{SamplingState}-{ParentSpanId}
+    headers["b3"] = f"{trace_id}-{span_id}"
+
+
+async def _on_request_start(
+    session: aiohttp.ClientSession,
+    context: types.SimpleNamespace,
+    params: aiohttp.TraceRequestStartParams,
+) -> None:
+    trace_ctx = context.trace_request_ctx
+    trace_id = getattr(trace_ctx, "trace_id", None)
+    if trace_id is None:
+        return
+    span_id = gen_trace_id()
+    _update_headers(params.headers, trace_id, span_id)
+
+
+def _make_trace_config() -> aiohttp.TraceConfig:
+    trace_config = aiohttp.TraceConfig()
+    trace_config.on_request_start.append(_on_request_start)  # type: ignore
```

Comment:
Why ignore?

Reply:
I suspect it is mypy bug, very close to your recent report:
```
neuromation/api/tracing.py:40: error: Argument 1 to "append" of "MutableSequence" has incompatible type "Callable[[ClientSession, SimpleNamespace, TraceRequestStartParams], Coroutine[Any, Any, None]]"; expected "Callable[[ClientSession, SimpleNamespace, <union: 15 items>], Awaitable[None]]"
```
Coroutine is not resolved as Awaitable or vice versa.